ros2node/ros2node/api/__init__.py

```diff
@@ -71,6 +72,102 @@ def get_node_names(*, node, include_hidden_nodes=False):
+            enclave=t[2])
+        for t in node_names_and_namespaces_with_enclaves
+        if (
+            include_hidden_nodes or
+            (t[0] and not t[0].startswith(HIDDEN_NODE_PREFIX))
+        )
+    ]
+
+
+def get_topics(remote_node_name, func, *, include_hidden_topics=False):
+    node = parse_node_name(remote_node_name)
+    names_and_types = func(node.name, node.namespace)
+    return [
+        TopicInfo(
+            name=t[0],
+            types=t[1])
+        for t in names_and_types if include_hidden_topics or not _is_hidden_name(t[0])]
+
+
+def get_subscriber_info(*, node, remote_node_name, include_hidden=False):
+    return get_topics(
+        remote_node_name,
+        node.get_subscriber_names_and_types_by_node,
+        include_hidden_topics=include_hidden
+    )
+
+
+def get_publisher_info(*, node, remote_node_name, include_hidden=False):
+    return get_topics(
+        remote_node_name,
+        node.get_publisher_names_and_types_by_node,
+        include_hidden_topics=include_hidden
+    )
+
+
+def get_service_client_info(*, node, remote_node_name, include_hidden=False):
+    return get_topics(
+        remote_node_name,
+        node.get_client_names_and_types_by_node,
+        include_hidden_topics=include_hidden
+    )
+
+
+def get_service_server_info(*, node, remote_node_name, include_hidden=False):
+    return get_topics(
+        remote_node_name,
+        node.get_service_names_and_types_by_node,
+        include_hidden_topics=include_hidden
+    )
+
+
+def get_action_server_info(*, node, remote_node_name, include_hidden=False):
+    remote_node = parse_node_name(remote_node_name)
+    names_and_types = node.get_action_server_names_and_types_by_node(
+        remote_node.name, remote_node.namespace)
+    return [
+        TopicInfo(
+            name=n,
+            types=t)
+        for n, t in names_and_types if include_hidden or not _is_hidden_name(n)]
+
+
+def get_action_client_info(*, node, remote_node_name, include_hidden=False):
+    remote_node = parse_node_name(remote_node_name)
+    names_and_types = node.get_action_client_names_and_types_by_node(
+        remote_node.name, remote_node.namespace)
+    return [
+        TopicInfo(
+            name=n,
+            types=t)
+        for n, t in names_and_types if include_hidden or not _is_hidden_name(n)]
+
+
+class NodeNameCompleter:
+    """Callable returning a list of node names."""
+
+    def __init__(self, *, include_hidden_nodes_key=None):
+        self.include_hidden_nodes_key = include_hidden_nodes_key
+
+    def __call__(self, prefix, parsed_args, **kwargs):
+        include_hidden_nodes = getattr(
+            parsed_args, self.include_hidden_nodes_key) \
+            if self.include_hidden_nodes_key else False
+        with NodeStrategy(parsed_args) as node:
+            return [
+                n.full_name for n in get_node_names(
+                    node=node, include_hidden_nodes=include_hidden_nodes)]
+
```

Comment:
why we see this difference?

Reply:
Sorry, I'm not sure how all that was duplicated. I'll remove it.